nav2_mppi_controller/include/nav2_mppi_controller/motion_models.hpp

```diff
@@ -70,12 +83,30 @@ class MotionModel
     // }
 
     const bool is_holo = isHolonomic();
+    float max_delta_vx = model_dt_ * control_constraints_.ax_max;
+    float min_delta_vx = model_dt_ * control_constraints_.ax_min;
+    float max_delta_vy = model_dt_ * control_constraints_.ay_max;
+    float max_delta_wz = model_dt_ * control_constraints_.az_max;
+    float vx_last = state.vx(0, 0);
+    float vy_last = state.vy(0, 0);
+    float wz_last = state.wz(0, 0);
     for (unsigned int i = 0; i != state.vx.shape(0); i++) {
       for (unsigned int j = 1; j != state.vx.shape(1); j++) {
-        state.vx(i, j) = state.cvx(i, j - 1);
-        state.wz(i, j) = state.cwz(i, j - 1);
+        float & cvx_curr = state.cvx(i, j - 1);
+        cvx_curr = std::clamp(cvx_curr, vx_last + min_delta_vx, vx_last + max_delta_vx);
```

Comment:
Oh. You're modifying the underlying `cvx` asset to be in the dynamic constraints. Change all the `float &` to a `float` so they you don't modify `cvx` and see if that fixes the issue. You shouldn't modify the underlying control requested, just the velocity output that the control represents after applying the motion model

Reply:
But in our case, the control output is the desired velocity, and `control_sequence` is calculated by performing softmax on `cvx`. So, if we don't constrain `cvx`, then it may result in acceleration crossing the bounds.

Reply:
for that, we will need to clamp the final control sequence as you said. But I'm not sure what will be the effect of applying hard constraints on the final control sequence.

Reply:
Well first - is this what fixes the acceleration problem? Or was it the other things that fixed it.

> So, if we don't constrain cvx, then it may result in out of bounds acceleration.

Ah yes. This is back to an old question I had: [should these be `cvx` or be `vx`?](https://github.com/ros-navigation/navigation2/blob/main/nav2_mppi_controller/src/optimizer.cpp#L378-L382) Previously it didn't matter since they were the same, but now they're not. I postulate that if the dynamic model applies is predictive (i.e. if we apply _this_ control, _this_ velocity occurs) then it should be `cvx`. But, if the dynamic model is applying restrictions, then I think it should be `vx` so we are using the restricted velocities.

> we will need to clamp the final control sequence as you said

We should do that irregardless. However, I think if we just clamp while using `cvx` that could create differences between the scored and optimal softmaxed trajectories. We should attempt to minimize or eliminate that difference as much as possible. Relying on that clamp should be dealing with small rounding errors, not major differentiations. 

So options...

- If changing the `float &` to `float` isn't what fixed your issues, then revert that? But I suspect that **is** what fixed your issues
- Change the softmax application to the velocities, not the controls
- Keep using the controls for the softmax but  add the acceleration constraints to the `ConstraintCritic` and score breaches negatively. That will minimize the breaches like we do with velocities so that when we do hit the `applyControlSequenceConstraints`, we're applying it to only rounding error levels. We have a recipe for that with the  velocities that works. 

It may be worth trying a couple of these to see what works the best

Reply:
> If changing the float & to float isn't what fixed your issues, then revert that? But I suspect that is what fixed your issues

It was due to that bug in setting vx_last. I've tried running with float& again, it reaches maximum velocity.
Do we have existing full integration tests for navigation through poses or waypoint following?

Reply:
> It was due to that bug in setting vx_last. 

Oh, ok... so it goes full speed again? Umm.. I think that works then!

Reply:
> Oh, ok... so it goes full speed again? Umm.. I think that works then! Can you show me the plots like you did before?

Yes sure!
![Screenshot from 2024-05-30 00-36-17](https://github.com/ros-navigation/navigation2/assets/64409716/924678ca-b1ec-4a25-b4f2-75728cc80ce7)

This is the global plan on which I'm testing
![Screenshot from 2024-05-28 01-59-52](https://github.com/ros-navigation/navigation2/assets/64409716/c6070435-2da8-4d07-96ca-0a74823e28d0)

Reply:
Oh ok! Well that works then I think!